### DISCUSSION TITLE
Module register nuxi command

### DISCUSSION THREAD

**Discussion by ricardogobbosouza** [nuxt/nuxt#17767] [Ideas]
It would be interesting for the module to be able to register commands in nuxi

```js
import { defineNuxtModule, defineNuxtCommand } from '@nuxt/kit'

export default defineNuxtModule({
  meta: {
    configKey: 'module:key'
  },

  setup (moduleOptions, nuxt) {
    defineNuxtCommand({
      meta: {
        name: 'command',
        usage: 'npx nuxi module:key command'
      },
      async invoke (args) {
        console.log(args)
      }
    })
  }
})
```

Ex:
```bash
npx nuxi module:key command
```

@pi0  do you think it's valid to create a PR for this?

**Comment by atinux** [comment]
Do you have an example in mind for a specific module?

**Reply by ricardogobbosouza** [reply]
The module could register commands to be executed.

router-module: Register a command `router list` that lists the app routes
google-fonts-module: Register a command `goole-fonts download` which will force download google fonts

**Reply by pi0** [reply]
> router-module: Register a command router list that lists the app routes

With comming [entity management commands](https://github.com/nuxt/framework/issues/3903), listing and modifing directory strcture will be possible out of the box

> google-fonts-module: Register a command goole-fonts download which will force download google fonts

Would this needed out of Build lifecycle? In this case, i would extend `nuxi prepare` with a hook to do preparation like downloading assets.

**Comment by Intevel** [comment]
What do you mean with this? What should the module register command do?

**Reply by ricardogobbosouza** [reply]
The module could register commands to be executed

**Comment by pi0** [comment]
Hi @ricardogobbosouza. Supporting external commands, is not possible within Nuxt modules since they are bound to Nuxt lifecycle and CLI starts way earlier. Instead, we supported new `nuxt-` command convention. As, if a module expose a binary called `nuxt-foo`, `nuxt foo` command will execute it. (please see https://github.com/nuxt/rfcs/issues/11 for more background).

We need to eventually add it to Nuxt 3 as well (https://github.com/nuxt/framework/issues/4060). Only blocker was to resolve one of main problems we had with Nuxt 2 due to lack of shared external util, a module CLI, had to implement wrapper and invoke for itself. (See [nuxt telemetry](https://github.com/nuxt/telemetry/blob/main/src/cli.ts) as an example)

**Reply by ricardogobbosouza** [reply]
@pi0 that's what i was referring to https://github.com/nuxt/module-builder/issues/32
